### PR TITLE
8267982: Set the node after peephole optimization to be removed

### PR DESCRIPTION
**Pull request by sunny868**
Hi all,

Enable peephole can cause SIGSEGV sometimes in OopFlow::build_oop_map(Node*, int, PhaseRegAlloc*, int*)+0x238 , so must set a removed-flags for the peephole-nodes , to avoid be set-live-bit() and cause crash.
 Please review this trivial change.

Thanks,
Sun Guoyun

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267982](https://bugs.openjdk.java.net/browse/JDK-8267982): Set the node after peephole optimization to be removed

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4276/head:pull/4276` \
`$ git checkout pull/4276`

Update a local copy of the PR: \
`$ git checkout pull/4276` \
`$ git pull https://git.openjdk.java.net/jdk pull/4276/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4276`

View PR using the GUI difftool: \
`$ git pr show -t 4276`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4276.diff">https://git.openjdk.java.net/jdk/pull/4276.diff</a>

</details>

### CODE REVIEW

**Comment by TobiHartmann** [comment]
The PR title needs to match the JIRA bug title (including the bug number). Please also change it to something more descriptive (like the one you had previously).

**Comment by sunny868** [comment]
@TobiHartmann what I need to do ?

**Comment by dean-long** [comment]
I think Tobias was suggesting that you change both this PR's title and the JBS issue back to "8267982: Set the node after peephole optimization to be removed"

**Review by vnkozlov — CHANGES_REQUESTED** [review]
To have new flag field in base class Node is expensive (size wise). I suggest to add it to  MachNode on which peephole is executed.
 
Otherwise the fix is correct.

File: src/hotspot/share/adlc/output_c.cpp [hunk 1]
```diff
@@ -1342,6 +1342,8 @@ static void generate_peepreplace( FILE *fp, FormDict &globals, PeepMatch *pmatch
     assert( false, "ShouldNotReachHere();");
   }
 
+  for( int i = 0; i <= max_position; i++)
+    fprintf(fp, "        inst%d->set_removed();\n", i);
```
Comment (vnkozlov) [inline src/hotspot/share/adlc/output_c.cpp:1346]: Code style. Use `{}`:
```
for () {
}
```

File: src/hotspot/share/opto/buildOopMap.cpp [hunk 2]
```diff
@@ -466,8 +466,8 @@ static void do_liveness(PhaseRegAlloc* regalloc, PhaseCFG* cfg, Block_List* work
           assert(def != 0, "input edge required");
           int first = regalloc->get_reg_first(def);
           int second = regalloc->get_reg_second(def);
-          if( OptoReg::is_valid(first) ) set_live_bit(tmp_live,first);
-          if( OptoReg::is_valid(second) ) set_live_bit(tmp_live,second);
+          if( OptoReg::is_valid(first) && !def->get_removed()) set_live_bit(tmp_live,first);
+          if( OptoReg::is_valid(second) && !def->get_removed()) set_live_bit(tmp_live,second);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/buildOopMap.cpp:470]: You need `def->as_Mach()->get_removed()` because `def ` is `Node*`.

Comment (sunny868) [reply src/hotspot/share/opto/buildOopMap.cpp]: Done

File: src/hotspot/share/adlc/output_c.cpp (hunk 1 above)
Comment (sunny868) [reply src/hotspot/share/adlc/output_c.cpp]: Done

**Comment by sunny868** [comment]
I made a mistake operation . I will fix it right away.

**Comment by sunny868** [comment]
@TobiHartmann @vnkozlov can you help me ? I made a mistake to closed this PR, but now I can not re-open this PR.
My new commit is  https://github.com/sunny868/jdk/commit/00a73ec58c9102132609765997b47b7f2c409833 .

**Comment by dean-long** [comment]
@sunny868 If you search for "reopen pull request after force push" you can find some helpful suggestions on how to fix this.  From what I read, you need to do a another force push back to the previous PR state before it can be reopened.  If all else fails, try opening a new PR.
Do we have any github/skara experts who can help with fixing this?

**Comment by sunny868** [comment]
@dean-long Thank you very much for your help ！
@vnkozlov bI had a new modify in buildOopMap.cpp . Please reviews it again,thanks.

File: src/hotspot/share/opto/buildOopMap.cpp [hunk 3]
```diff
@@ -466,8 +466,12 @@ static void do_liveness(PhaseRegAlloc* regalloc, PhaseCFG* cfg, Block_List* work
...
           int first = regalloc->get_reg_first(def);
           int second = regalloc->get_reg_second(def);
-          if( OptoReg::is_valid(first) ) set_live_bit(tmp_live,first);
-          if( OptoReg::is_valid(second) ) set_live_bit(tmp_live,second);
+          if ( def->is_Mach() && def->as_Mach()->get_removed() ) {
+            //do nothing. peephole has removed the node.
+          } else {
+            if( OptoReg::is_valid(first) ) set_live_bit(tmp_live,first);
+            if( OptoReg::is_valid(second) ) set_live_bit(tmp_live,second);
+          }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/buildOopMap.cpp:474]: Code style. Use next pattern in code you modified (body on separate line and no spaces before/after condition):
```
if (cond) {
  foo;
}
``` 
We want to correct style in all places eventually as we work on code.

**Review by vnkozlov — COMMENTED** [review]
Looks good. I have only style fix comment.
I will test changes meanwhile because I don't see it passed through GitHub testing.

File: src/hotspot/share/opto/buildOopMap.cpp (hunk 3 above)
Comment (sunny868) [reply src/hotspot/share/opto/buildOopMap.cpp]: Done

**Review by vnkozlov — APPROVED** [review]
Good.
Tier1-3 testing passed clean.

**Comment by sunny868** [comment]
@vnkozlov  Could you please sponsor it for me?

File: src/hotspot/share/opto/machnode.hpp [hunk 4]
```diff
@@ -373,6 +376,8 @@ class MachNode : public Node {
 
   // Returns true if this node is a check that can be implemented with a trap.
   virtual bool is_TrapBasedCheckNode() const { return false; }
+  void set_removed(){ _removed = true; }
+  bool get_removed(){ return _removed; }
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/machnode.hpp:380]: Please add whitespaces:
`set_removed() {`
`get_removed() {`

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Otherwise, it looks good to me.

File: src/hotspot/share/opto/machnode.hpp (hunk 4 above)
Comment (sunny868) [reply src/hotspot/share/opto/machnode.hpp]: Done

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Comment by sunny868** [comment]
@vnkozlov @TobiHartmann Could you please sponsor it for me?

**Pull request closed without merging**